README.md

*StreamHelper* is a program to centralize and simplify the creation and updating of textfiles with set names.

It is mainly intended for small to semi-professional streamers, who need to control a rather large amount of textfiles while streaming (e. g. OBS, Streamlabs, etc.)


#### These third party libraries are being used by the program:

 -  [PyQt5 Gui, Widgets, Core](https://pypi.org/project/PyQt5/)
 - [pyperclip](https://pypi.org/project/pyperclip/)
 
StreamHelper is implemented in **python** and is being distributed with the help of  [pyinstaller](https://www.pyinstaller.org/) and [innoSetup](https://jrsoftware.org/isinfo.php) 

## Installation
Since this is *not* a signed installer, Windows Smart Screen will warn you on installation. This is normal and perfectly fine and cannot be circumvented at the moment. 

GitHub download speeds are sometimes pretty low. To ease access to the installer, I also uploaded the file on mega:
https://mega.nz/folder/JOxmABTQ#Dn8tYTcBoKMxezW61TWK0A

1. Open the installer by double-clicking
2. Select, if you want to install the program for *all users (recommended)* or just the current one
3. Accept the license and set the path
4. Read the FAQ and the changelog
5. Start the  program from the location or your desktop

0.3.1 has not been packed into an installer. If you want to use it, download from mega, extract where you want to and run.

There shouldn't be any issues, since this process is straightforward.
*Note (Pre-0.3.1): StreamHelper can only be started from the wizard if you gave it elevated privileges, otherwise it will fail. This is expected, since the program is only able to run with elevation!*

## Startup
On startup, a folder will be created at *C:\Users\\<username\>\AppData\Local\StreamHelper* and all basic files will get created. This includes:

- The  autosave
	
- The config file
		
- The folder *textfiles* with all subfolders *Lists*, *eSports* and *API (unused in 0.3)*

The program on any version before 0.3.1 will *not* start if you don't have elevated privileges! 
From this version onwards, the program will ask for these privileges as soon as it needs it.

## Usage
Using this program is straightforward and rather simple, there are tooltips for most buttons and sections. However, there are some things to know before starting to produce with it:

- The eSports tab will auto-update every line edit to the corresponding text file when its focus is lost (if you click somewhere else), to save the contents to the autosave though, you will have to use the "save" button on each section
- The "auto-update lists" setting is only applicable for list items and will update them as soon as you click somewhere else. If it is turned off, the files will only update if you manually press the "update" button. To autosave them, you will still need to use the "update" button regardless of the setting.
- Resetting the settings only takes full effect on a restart of the application
- Swapping two items/contents automatically updates the corresponding textfiles.
- Naming the lists only has an effect if "List-Split" is activated.
- You can export settings by copying your "config.json" somewhere else. If you move it instead your settings get reset.
- Deleting folders manually while using the program might cause StreamHelper to crash (known bug)
- Keyboard shortcuts "ctrl+l, ctrl+v" for loading of files is deactivated (known bug)

A complete list of changes, documentation and bugs is in the works, however, since interest on this project is low at best, these will take some time.

More will be added to this list until I have a proper documentation with tutorials. Currently though, since the application isn't too big, please bear with me if I focus on programming.

If you happen to have any questions regarding usage of this program, feel free to ask me, write me an e-mail or dm me on Discord (Witheria #4806)!

## Deinstallation
You can deinstall StreamHelper from the "installed programs" section within your PC's settings. 

0.3.1: Simply delete the directory you decided to put the files and the appdata folder.


### Thank you for being a part of this project!


[Donate...](https://www.paypal.com/donate?hosted_button_id=825RPFTRDCW7A)




    © Toni Schmidbauer 2020 / Released under GPL v3
 
    Bugs, problems, critique: schmidbauer.biz@gmail.com


resources/CHANGELOG.txt
- load multiple oi files
- selective loading of oi files
- load oi files from server




************************************************************************************************************************


OPEN ALPHA 0.3.1


Added:
    - Chrono Items (Counting up, down and showing time)
    - Image Items (Copying images to the standard path)
    - eSports fields "Infotext", "Caster" and "Custom" in the "More..." field with funneling
    - Some additional settings to fit your needs
    - selective swapping, moving and deleting (im working on drag/drop! Its hard work ok!!!)
        -> just click on the items you want to move/swap/delete

- The UI has been stylized with frames. I am not putting enough work into UI design, but oh well, it works

- The Documentation is in work, but since it will be a lot of work, I'll only be posting updates whenever it
  fits my flow. Since noone else is currently really interested in this project yet, it doesn't really matter.

Bugs fixed:
    - The application does not crash anymore on the follow occasions
        - The config.json gets deleted while running
        - The config.json gets modified while running
        - The config.json gets modified to an unreadable format while closed

        - same for autosave.json

    - The chrono format does not randomly get mixed up anymore
    - There is now a custom format setting on each chrono item
    - Chrono items have now icons instead of text

    - Image items don't feel like crashing anymore if the image they are attached to gets deleted
    - Image items can now actually be swapped
    - Image items are properly saved even after resetting the lists

    - A mass of bugs while development have been fixed, but there might be still many left.
    Help would be appreciated, feel free to contact me

************************************************************************************************************************
OPEN ALPHA 0.3.0


Added:
    - Full eSports UI functionality
    - List item buttons connectivity (except "change")
    - Icons to make it look nicer (credits can be found under the menu item "help")
    - Settings are working now. Please send your suggestions to the dev
    - About tab in the settings displays information
    - Save/Load function is also working for the eSports UI
    - Just try it out, it's the best for finding bugs which you can then send me (hehe)

Removed:
    - Chrono Item (still in development)
    - Mask functions aren't taking space away now
    - Other unnecessary code snippets and garbage
    - API functionality (sad League noises)

Will be probably implemented in the next update:
    - Chrono Items (Timers)
    - Teams implementation
    - More customization and settings
    - Drag enabled lists (You will be able to drag items into the other list!)

On the TODO-List:
    - Drag and Drop interactions with files and links
    - APIs (Toornament, LoL, RocketLeague, Overwatch)
    - Team Tree
    - More settings
    - More design options
    - Upgrade to PyQt6 (well, far, far away)


Hear me out, this was a disaster to program. It took so much time I even had to make a whole update dedicated to that
(see 0.2.2). I believe I'm at around 250 work hours into this, so, can't quit now I guess.

The JSON stuff is working perfectly, I'm keeping it and with that the program simple.

Log functionality is still in development, I will probably need to restructure a lot of this stuff to make it work
properly. Exceptions aren't caught yet, start the program from the command line to see them. A lot is still just "print".

************************************************************************************************************************
CLOSED ALPHA 0.2.2


Added:
    - Save/Load function (also Autosave works again)
    - The Lists are back up and working again
    - eSports UI (Not connected yet)
    - API functionality (retrieve summoner name from client)


This update was kind of heavy, I rewrote the whole textlists idea. Formerly split into several methods,
I subclassed the QListWidgetItem class into the elements I'll be using.
It will be interesting to see where this road leads to.

Also, I will not be using XML formatted data in my files anymore. JSON proved to be faster and more flexible,
so that's my format of choice now.

It took way too much time to get everything back up and running again,
but well, it is much less text now ( getTextofItem() was 147 lines, now its like 50 ).

************************************************************************************************************************
CLOSED ALPHA 0.2.1


Added:
    - Settings window
    - Settings UI
    - Config File functionality

Temporarily REMOVED:
    - Save/Load function
    - Text and Number widgets

************************************************************************************************************************
CLOSED ALPHA 0.2.0


Added:
    - The LoL Mask to have everything neatly fit in with your gaming needs
    - Full mask functionality
    - Configs can now be imported and exported
    - A "Settings" Menu Option for your convenience

+++ A guide and startup manual will be written and released soon, even though there aren't many features yet +++

APIs are not working yet. This will come in a later update

I have also distanced myself from the "Mask" idea that I initially had ( import mask when you need it,
remove if you dont ), because it is simply too much work to implement it. So, the package will just include everything
and you WILL BE ABLE to remove some UIs/tabs in the settings later on. They wont get fully deleted, but wont
load on startup, so that's a plus i guess.

I split all the code in several different files for organizing purposes. Feel free to comment on this, I thought it
would be better than having everything in like three files.

************************************************************************************************************************

resources/runtime/savestate.py
import os
from inspect import getsourcefile

from PyQt5.QtCore import QTimer

from sys import platform

from pathlib import Path

home = str(Path.home())

print("Platform is " + platform, ", Home directory is " + home)
standardFilePath = "None"
symbol = "None"
if platform == "linux":
    try:
        os.mkdir(home + "/Documents/StreamHelper")
    except FileExistsError:
        print("\n")
    standardFilePath: str = home + "/Documents/StreamHelper"
    symbol: str = "/"
    system = 0
elif platform == "win64":
    standardFilePath: str = os.getenv('LOCALAPPDATA') + "\\StreamHelper"
    symbol: str = "\\"
    system = 0
elif platform == "win32":
    standardFilePath: str = os.getenv('LOCALAPPDATA') + "\\StreamHelper"
    symbol: str = "\\"
    system = 0

# Installation path, app id, product info
SOURCE_PATH: str = os.path.split(os.path.abspath(getsourcefile(lambda: 0)))[0].replace("runtime", "")
APP_ID: str = "ToniSchmidbauer.StreamHelper.StreamHelperOpenAlpha.0.3.1b"
PRODUCT_VERSION: int = 0o0003102

# Standard Filepath and File Names
# standardFilePath: str = os.getcwd()
# standardFileNames = {"chronoup", "chronodown", "time"} UNUSED
standardXMLNames = ['autosave', 'config', 'settings']
standardDirNames = ['Lists', 'eSports', 'API']
forbiddenChars = ["/", ">", "<", ":", "\"", "\\", "|", "?", "*", "CON", "PRN", "AUX", "NUL", "COM1", "COM2", "COM3",
                  "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
                  "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"]

# Saving lists and values
saveLists: dict = {"Left": None, "Right": None}
saveListItems: dict = {"Left": {}, "Right": {}}
saveListData: dict = {"Left": {}, "Right": {}}
saveListItemNames: list = []
yeOldeSaveToCompare: dict = {"Left": {}, "Right": {}}  # This exists to compare different states of the program.

# create all necessary lists
txtlist = {"CS_Day": "", "CS_Game": "", "CS_Group": "", "CS_League": "", "CS_Tournament": "",
           "HT_Player1": "", "HT_Player2": "", "HT_Player3": "", "HT_Player4": "", "HT_Player5": "",
           "HT_Sub1": "", "HT_Sub2": "", "HT_abbreviation": "", "HT_city": "", "HT_Comment": "",
           "HT_name": "", "HT_organisation": "",
           "T1_name": "", "T1_score": "", "T1_city": "",
           "T2_name": "", "T2_score": "", "T2_city": "",
           "T3_name": "", "T3_score": "", "T3_city": "",
           "T4_name": "", "T4_score": "", "T4_city": ""
           }
morelist = {"Caster1": "",
            "Caster2": "",
            "Caster3": "",
            "Stat1": "",
            "Stat2": "",
            "Stat3": "",
            "Custom1": "",
            "Custom2": "",
            "Custom3": "",
            }
folderlist = ["Competitive Streaming", "Home Team", "Teams", "More fields"]

lineedit_list = {}

# Standard autosave look
autosave_standard = {"Lists": saveListData, "eSports": {"txtlist": txtlist, "morelist": morelist}}

# Setting settings
# Look for images at these locations (they have to be named according to the guidelines)
imageLocations = {"0": standardFilePath + symbol + "images"}

# Image names
imageNames = {

}

# On changing list names, this will process the name so the old folder gets deleted
NewListName = tuple()

# Dict to process the settings tree and connect it to the stacked widget
settingsList = {
    "Main Settings": 0,
    "Riot API": 0,
    "File Paths": 1,
    "Lists": 2,
    "Miscellaneous": 3,
    "Riot Settings": 4,
    "Statistics": 5,
    "Timers": 6,
    "About": 6
}
# This list will be loaded on startup from the settings file, and saved on request or shutdown. Any changes will
# therefore apply immediately
configList = {
    # Basic settings
    "StartupTab": 0,
    "StartupSettingsTab": 0,
    "Product Version": PRODUCT_VERSION,

    # File path settings
    "UserFilePath": 0,
    "CustomFilePath": standardFilePath,
    "ImagePaths": imageLocations,

    # List settings
    "AllowedItems": {
        "Chrono Item": True,
        "GT Item": False,
        "Number Item": True,
        "Text Item": True,
        "Image Item": True
    },
    "GlobalPretext": "",
    "ListSplit": False,
    "AutoUpdateFiles": False,
    "RightListName": "Right",
    "LeftListName": "Left",

    # eSport Settings
    "funnelfile_separator": "\n",

    # Riot API settings
    "FontSize": 11,
    "ChronoFormat": "hh:mm:ss",

    # EKO API settings
    "serverkey": "",
    "serveruri": "",
    "opt_params": "Limit=1",
    "ign_params": "",
    "srv_file_name": "server.txt",
    "srv_interval": 12,
    "svr_sep": "\n"
}
# List used for resetting the settings (wont be changed around and read like the other one)
standardConfigList = {
    "WARNING": "EVERY MANUAL CHANGE REQUIRES A RESTART OF THE PROGRAM!",
    # Basic settings
    "StartupTab": 0,
    "StartupSettingsTab": 0,

    # File path settings
    "UserFilePath": 0,
    "CustomFilePath": standardFilePath,
    "ImagePaths": imageLocations,

    # List settings
    "AllowedItems": {
        "Chrono Item": True,
        "GT Item": False,
        "Number Item": True,
        "Text Item": True,
        "Image Item": True
    },
    "GlobalPretext": "",
    "ListSplit": False,
    "AutoUpdateFiles": False,
    "RightListName": "Right",
    "LeftListName": "Left",

    # eSport Settings
    "funnelfile_separator": "\n",

    # Riot API settings
    "FontSize": 11,
    "ChronoFormat": "hh:mm:ss",

    # EKO API settings
    "serverkey": "",
    "serveruri": "",
    "opt_params": "Limit=1",
    "ign_params": "",
    "srv_file_name": "server.txt",
    "srv_interval": 12
}

# This integer counts how many times the function saveChronoToAutosave in textlists.program has been accessed
chrono_access_counter: int = 0

# This saves the currently active Summoner for the session
summoner = ""

# Stuff needed for eSports

# This exists to allow the score operations to work
lastScoreChange = 0

# Since there are a lot of access requests to readAutosave(), only every %s request will get logged
AUTOSAVE_LOG_LEVEL_OUTPUT = 4

# This is the timer to clock the chronoitems
timer: QTimer = None

# Create the bonus fields window and connect it
ExtraFieldWidget = None

# important mappings for the League API
champIdMaps: dict = {
    "266": "Aatrox",
    "103": "Ahri",
    "84": "Akali",
    "12": "Alistar",
    "32": "Amumu",
    "34": "Anivia",
    "1": "Annie",
    "523": "Aphelios",
    "22": "Ashe",
    "136": "AurelionSol",
    "268": "Azir",
    "432": "Bard",
    "53": "Blitzcrank",
    "63": "Brand",
    "201": "Braum",
    "51": "Caitlyn",
    "164": "Camille",
    "69": "Cassiopeia",
    "31": "Chogath",
    "42": "Corki",
    "122": "Darius",
    "131": "Diana",
    "119": "Draven",
    "36": "DrMundo",
    "245": "Ekko",
    "60": "Elise",
    "28": "Evelynn",
    "81": "Ezreal",
    "9": "Fiddlesticks",
    "114": "Fiora",
    "105": "Fizz",
    "3": "Galio",
    "41": "Gangplank",
    "86": "Garen",
    "150": "Gnar",
    "79": "Gragas",
    "104": "Graves",
    "120": "Hecarim",
    "74": "Heimerdinger",
    "420": "Illaoi",
    "39": "Irelia",
    "427": "Ivern",
    "40": "Janna",
    "59": "JarvanIV",
    "24": "Jax",
    "126": "Jayce",
    "202": "Jhin",
    "222": "Jinx",
    "145": "Kaisa",
    "429": "Kalista",
    "43": "Karma",
    "30": "Karthus",
    "38": "Kassadin",
    "55": "Katarina",
    "10": "Kayle",
    "141": "Kayn",
    "85": "Kennen",
    "121": "Khazix",
    "203": "Kindred",
    "240": "Kled",
    "96": "KogMaw",
    "7": "Leblanc",
    "64": "LeeSin",
    "89": "Leona",
    "876": "Lillia",
    "127": "Lissandra",
    "236": "Lucian",
    "117": "Lulu",
    "99": "Lux",
    "54": "Malphite",
    "90": "Malzahar",
    "57": "Maokai",
    "11": "MasterYi",
    "21": "MissFortune",
    "62": "MonkeyKing",
    "82": "Mordekaiser",
    "25": "Morgana",
    "267": "Nami",
    "75": "Nasus",
    "111": "Nautilus",
    "518": "Neeko",
    "76": "Nidalee",
    "56": "Nocturne",
    "20": "Nunu",
    "2": "Olaf",
    "61": "Orianna",
    "516": "Ornn",
    "80": "Pantheon",
    "78": "Poppy",
    "555": "Pyke",
    "246": "Qiyana",
    "133": "Quinn",
    "497": "Rakan",
    "33": "Rammus",
    "421": "RekSai",
    "526": "Rell",
    "58": "Renekton",
    "107": "Rengar",
    "92": "Riven",
    "68": "Rumble",
    "13": "Ryze",
    "360": "Samira",
    "113": "Sejuani",
    "235": "Senna",
    "147": "Seraphine",
    "875": "Sett",
    "35": "Shaco",
    "98": "Shen",
    "102": "Shyvana",
    "27": "Singed",
    "14": "Sion",
    "15": "Sivir",
    "72": "Skarner",
    "37": "Sona",
    "16": "Soraka",
    "50": "Swain",
    "517": "Sylas",
    "134": "Syndra",
    "223": "TahmKench",
    "163": "Taliyah",
    "91": "Talon",
    "44": "Taric",
    "17": "Teemo",
    "412": "Thresh",
    "18": "Tristana",
    "48": "Trundle",
    "23": "Tryndamere",
    "4": "TwistedFate",
    "29": "Twitch",
    "77": "Udyr",
    "6": "Urgot",
    "110": "Varus",
    "67": "Vayne",
    "45": "Veigar",
    "161": "Velkoz",
    "254": "Vi",
    "234": "Viego",
    "112": "Viktor",
    "8": "Vladimir",
    "106": "Volibear",
    "19": "Warwick",
    "498": "Xayah",
    "101": "Xerath",
    "5": "XinZhao",
    "157": "Yasuo",
    "777": "Yone",
    "83": "Yorick",
    "350": "Yuumi",
    "154": "Zac",
    "238": "Zed",
    "115": "Ziggs",
    "26": "Zilean",
    "142": "Zoe",
    "143": "Zyra"
}


class LOCK:
    COMMON: dict = {
        'grant_type': 'authorization_code',
        'client_id': "",
        'client_secret': "",
        'redirect_uri': "",
        'code': ""
    }
    TOKEN: str = ""
    REFRESH: str = ""
    EXPIRE: int = 0
    TIME: int = 0
    URL: str = "https://www.streamlabs.com/api/v1.0/token"
